src/bin/julialauncher.rs

```diff
@@ -300,6 +302,181 @@ fn get_override_channel(
+                        break;
+                    } else {
+                        // `j == arg.len() - 1`
+                        skip_next = true;
+                    }
+                }
+            } else {
+                program_file = Some(arg.clone());
+                break;
+            }
+        }
+        if let Some(program_file) = program_file {
+            let mut path = PathBuf::from(program_file);
+            path.pop();
+            while !path.join("Project.toml").exists() && !path.join("JuliaProject.toml").exists() {
+                if !path.pop() {
+                    return None;
+                }
+            }
+            return Some(path);
+        } else {
+            return None;
+        }
+    } else if project.starts_with('@') {
+        let depot = match std::env::var("JULIA_DEPOT_PATH") {
+            Ok(val) => match val.split(':').next() {
+                Some(p) => PathBuf::from(p),
+                None => dirs::home_dir().unwrap().join(".julia"),
+            },
+            _ => dirs::home_dir().unwrap().join(".julia"),
+        };
+        let path = depot.join("environments").join(&project[1..]);
+        if path.exists() {
+            return Some(path);
+        } else {
+            return None;
+        }
+    } else {
+        return Some(PathBuf::from(project));
+    }
+}
+
+fn julia_version_from_manifest(path: PathBuf) -> Option<Version> {
+    let manifest = if path.join("Manifest.toml").exists() {
+        path.join("Manifest.toml")
+    } else if path.join("JuliaManifest.toml").exists() {
+        path.join("JuliaManifest.toml")
+    } else {
+        return None;
+    };
+    let file = std::fs::File::open(manifest).ok()?;
+    let reader = std::io::BufReader::new(file);
+    // This is a somewhat bootleg way to parse the manifest,
+    // but since we know the format it should be fine.
+    for line in reader.lines() {
+        let line = line.ok()?;
+        if line.starts_with("julia_version = ") {
+            return Version::parse(line["julia_version = ".len()..].trim_matches('"')).ok();
+        }
+    }
+    return None;
+}
+
+fn get_julia_path_for_version(
+    config_data: &JuliaupConfig,
+    juliaupconfig_path: &Path,
+    version: &Version,
+) -> Result<PathBuf> {
+    let mut best_match: Option<(&String, Version)> = None;
+    for (installed_version_str, path) in &config_data.installed_versions {
+        if let Ok(installed_semver) = Version::parse(installed_version_str) {
+            if installed_semver.major != version.major || installed_semver.minor != version.minor {
+                continue;
+            }
+            if let Some((_, ref best_version)) = best_match {
+                if installed_semver > *best_version {
+                    best_match = Some((&path.path, installed_semver));
+                }
+            } else {
+                best_match = Some((&path.path, installed_semver));
+            }
+        }
+    }
```

Comment:
I think we should broadly use a different strategy here. Two problems:
1. I don't think we should just auto-handle patch updates. If I get a repo that records it uses Julia 1.10.1, and I only have 1.10.2 installed, I would want it to _not_ automatically use that. There might have been a really good reason that the Julia version was recorded as 1.10.1. We should think about lots of options to alert me to what is going on and make it easy to opt-in, but the default IMO should be what was recorded in the manifest, just how we do it for packages.
2. This doesn't handle platforms well. We can always have multiple versions installed for the same version that are only differentiated by platform (say `1.10.1~x64` and `1.10.1~x86`).

Here is how I would handle this instead:
1. Read the Julia version from the manifest
2. Look in the version db for the channel that matches that version exactly, and extract the exact Julia version from that (this makes for a 1:1 mapping from Julia semver to _one_ thing that can show up in `installed_versions`
3. Check whether that version exists in `installed_versions`, and then either use that or error (or in the future offer to install that channel).

Reply:
My handling of patch versions was informed by a conversation I was sitting in on in Triage a while back. I may be misremembering, but I recall discussion on how auto-selection of Julia version should interact with patch versions, and the conclusion was that the latest patch version should be used.

There are two aspects to this that I see:
- Whether you actually want bugs from an earlier Julia version
- Proliferation of installed Julia versions

In my mind, both of these point towards wanting to use the latest available patch version. The sole exception in my mind is when you *want* to reproduce bugs in the Base Julia, but I'm not sure that should be the guiding case for this design.

Having the same Julia version compiled for multiple platforms installed is an annoying edge-case...

Reply:
In my mind the entire point of a manifest is to _exactly_ pin down dependencies, so that things just run if one instantiates it. I see no reason why Julia itself should be handled differently here from any other package, after all we are not also suggesting that the package loader can just load newer patch versions of packages, right? I really want to get to a point where if I receive a manifest, I can be pretty sure that the person that created this manifest ran as close as possible the same versions of everything that will run on my machine if I instantiate it.

This would also lead to really weird UX problems. Say I download a replication archive from a paper that records Julia 1.10.0 in the manifest. That replication code works around a bug in 1.10.0. The bug has since been fixed in 1.10.1, but the replication code actually depends on the presence of the bug, so it won't run correctly under 1.10.1. As a user of that code, what am I supposed to do now? Uninstall 1.10.1 entirely from my system to force Juliaup to actually run with the version that is recorded in the manifest? I will now also get a different version of Julia when running depending on (completely unrelated) channels I have installed, that also seems really weird.

The worry about binary bloat in terms of installed Julia versions of course is real. But I think we should try to fix that via UX. For example, the launcher could print a notice when it starts a project, and the compat section in the `Project.toml` suggests that a newer version of Julia could be used for that project. It could even offer an interactive option "hey, you are running with a project that is using an outdated Julia version" or something like that.

(As a side note, I just checked, my `.julia` folder is at 27GB. If we want to fight binary bloat, there seems some other villain in town ;)

Reply:
Exact replication has occurred to me, and that is somewhat convincing.

However, in most projects I have manifests for, I want to use the newest patch version available. I don't think it should take as much effort to upgrade the patch version as the minor version, and I wouldn't see exact reproducibility at the expense of more involved patch version upgrading as a clear win.

Re: binary bloat, yes. I think we can manage that better via UX (I actually have a particular idea on this: I think juliaup should auto-install and GC them after some number of days of not being used, leaving manually added channels alone), but *also* installing less is good IMO (also from the UX perspective of waiting for more Julia versions to be installed less often).

Reply:
In my mind replication is such a central and important goal in scientific work that we should really support that well. Scientific code is, after all, an important target for Julia. And for that scenario, any of these fully auto-update mechanisms are really not good.

We should probably have configuration for some of this. We could have a config option `auto_update_manifests`. The default could be unconfigured. When the launcher is started with a manifest with an outdated Julia version, it could show a prompt `The Julia version in this project is outdated, do you want update it (u), run with the old version (n), or always update Julia versions in this case (a), never update Julia versions in this case (x)?`

In all of these cases we'll have to do a lot of checking, though, or probably will even have to run the resolver. At the end of the day, any of the packages listed in the manifest might actually exclude the newer Julia version via a compat section in its `Project.toml`, and we'll have to respect that. I think that is one of the main reasons that in my mind we should only ever run with a different Julia version if the package manager logic has signed off on that combination. But checking all these compats is not really something we can do on the hot-path of the launcher... 

>  I think juliaup should auto-install and GC them after some number of days of not being used, leaving manually added channels alone

Yes, agreed, I had thought about a similar idea. I think we should keep track of user projects in the config file, and then having that project tracked in the config file is the reason why that specific Julia version is not GCed. So essentially we would _not_ install a channel to instantiate a project, but instead just install that version as a dependency of the project.

Reply:
I'm very much on board with having exact replication be first-class.

I also think that most of the time, when a patch release comes out fixing bugs, that you want to use that.

So, ideally what I think we want to do here is to "have our cake and eat it too" by having both replication and patch-upgrades be easy. I'm not sure how best to do so.

Perhaps a prompt or a central list of per-project policies is the best way to do things?

Reply:
Yeah, I like the prompt idea best. So, I think essentially I would use the logic I outlined above to decide which Julia to launch, but augment it with logic that decides whether to show a "This project could be updated to a new Julia version" prompt.

I would propose that we implement that for this PR, and then the next step (but could be a new PR) could be that we detect interactive run scenarios, and in those show a prompt that offers to update the project.

> a central list of per-project policies

We might be able to avoid that if we lean on the `julia` entry in the `[compat]` section of `Project.toml`. Essentially the launcher could only show this update prompt for a project if that actually makes sense given the `julia` compat entry in the `Project.toml`. As a project author I could then completely prevent the prompt from showing by using `julia = "=1.10.3"` for example. Or if you want patch to be prompted about patch version updates, use `julia = "~1.10"`.

Reply:
Using the `[compat]` of the `Project.toml` for project/compatability intent of Julia seems like the obvious thing with the current tools we have. I like it!

That seems to strike a good balance of concerns to me, I'm happy to go ahead with this :+1:

Reply:
I pushed a commit that implements my original proposal of how to decide which binary to launch.